Add FNet Encoder Layer

abheesht17 opened the pull request:
Resolves #40. 


I've implemented the FNet Encoder Layer. I'll add unit tests once the code has been finalised. @mattdangerw ,@chenmoneygithub, please go through it. Thanks!

Here is the link to the Colab notebook where I do a forward pass: https://colab.research.google.com/drive/1ldWMSqw8q8qwndHUxtuILxowRRFiBTbT?usp=sharing 

I'm planning to add an example for this (text classification). I'll stack 2-3 encoder layers and train the model.
**Caveat:** In the paper, the authors mention that they used SentencePiece Tokenizer (this hasn't been implemented in `keras-nlp` yet). So, should I add the example later in a separate PR? Or should I just use SentencePiece from TF Text?

abheesht17 commented:
Hey, @mattdangerw. Any thoughts on this?

chenmoneygithub commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,157 @@
+        self._output_dropout = keras.layers.Dropout(rate=self.dropout)
+
+    def _fourier_transform(self, input):
+        # Apply FFT on the input and take the real part. Before we apply fourier
```
Please split this to two comments.

abheesht17 replied:
Done 👍🏼

chenmoneygithub commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,157 @@
+        if not self._built:
+            self._build(inputs.shape)
+
+        # Apply fourier transform on the input. Note: We don't have padding
```
I want to understand this a bit more - are you saying the original paper does not do padding, and take in RaggedTensor as input? Thanks!

abheesht17 replied:
So, as far as I understand, they do pad the sequences so that every sequence is of the same length. See this: https://github.com/google-research/google-research/blob/master/f_net/layers.py#L106. The expected shape is `(bsz, max_seq_length, hidden_dim)`.

However, the padding mask is not used at all. They delete it here: https://github.com/google-research/google-research/blob/master/f_net/layers.py#L114. So, the mixing is done over all the tokens, including the padding tokens. See the input pipeline here: https://github.com/google-research/google-research/blob/master/f_net/input_pipeline.py#L100.

abheesht17 replied:
I received a reply from the first author, James Lee-Thorp. This is the gist of the message:

**For FNet, padding tokens are applied, but padding tokens are not masked**. There is a possibility that the model may become sensitive to samples with varying number of padding tokens...but in practice, the authors did not notice any major performance issues. So, I think we can keep it the same, we don't need to apply padding masks.

If we do want to apply a padding mask, we should only apply it after the embedding layer, and not after every Fourier sublayer. This is because if we apply it after every Fourier sublayer, we will end up zeroing out some frequencies which is counter-productive.

chenmoneygithub reviewed (commented):
Thanks for the PR! 

Dropped some comments, and please add unit tests to test the layer's funcitonality, thanks!

mattdangerw commented:
Actually, if it's ok, maybe wait until this Tuesday to fully flesh this out? We have an internal sync meeting then, and will discussing where we want to set our bar (e.g. with citations) for inclusion in this repo.

I'm pretty sure we will want this, but if you wait till then, we could give a more confident green light. Thanks for patience! Really do appreciate the contributions!

abheesht17 commented:
Sure! Thank you :)

abheesht17 commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,157 @@
+    Args:
+        intermediate_dim: int, defaults to 3072. The hidden size of feedforward
+            network.
+        dropout: float, defaults to 0.1. the dropout value, applied in the
```
Done 👍🏼

abheesht17 reviewed (commented):
Addressed first round of comments.

abheesht17 commented:
Awesome, @mattdangerw! Thank you :). I'll add the the tests and the initialiser arguments. 

One more thing - could you and @chenmoneygithub please give https://github.com/keras-team/keras-nlp/pull/43#discussion_r830591662 and https://github.com/keras-team/keras-nlp/pull/43#discussion_r831721727 a look so that we can have closure on the padding mask issue?

abheesht17 commented:
@mattdangerw, @chenmoneygithub, I've done the following:
- [x] Added UTs (followed the UTs for TransformerEncoder).
- [x] Added kernel and bias initialisers.

Let me know if it's okay.

fchollet commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,183 @@
+        self.bias_initializer = keras.initializers.get(bias_initializer)
+        self._built = False
+
+    def _build(self, input_shape):
```
This should just be the built-in `build()`

abheesht17 replied:
Ah, overriding the `build()` function in `tf.keras.layers.Layer`? Should I make the same changes to the `TransformerEncoder` class? 
https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/transformer_encoder.py#L83

chenmoneygithub replied:
the default `build()` only works if your `call()` method takes one input, the TransformerEncoder takes multiple inputs, so default `build()` cannot be used there.

abheesht17 replied:
Ohh, I see. Thanks! :)

fchollet commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,183 @@
+        )
+        self._output_dropout = keras.layers.Dropout(rate=self.dropout)
+
+    def _fourier_transform(self, input):
```
In-line it in `call`

abheesht17 replied:
Done!

fchollet commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,183 @@
+        mixing_output = tf.math.real(tf.signal.fft2d(input))
+        return mixing_output
+
+    def _add_and_norm(self, input1, input2, norm_layer):
```
In-line it in `call`

abheesht17 replied:
Done!

fchollet commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,183 @@
+    def _add_and_norm(self, input1, input2, norm_layer):
+        return norm_layer(input1 + input2)
+
+    def _feed_forward(self, input):
```
In-line it in `call`

abheesht17 replied:
Done!

fchollet commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,183 @@
+
+    def __init__(
+        self,
+        intermediate_dim=3072,
```
Where does this value come from?

mattdangerw replied:
Let's remove the default. intermediate_dim should be a required argument to pass in.

chenmoneygithub replied:
At component level let's do not set the default, we will pass this value when we are building the actual model.

abheesht17 replied:
Ah, alright. I've removed the default value!

fchollet commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,183 @@
+    def __init__(
+        self,
+        intermediate_dim=3072,
+        dropout=0.1,
```
Should default to 0

mattdangerw replied:
Right, but we are not shipping a model architecture here, we are shipping a reusable layer. I would be inclined to use the same defaults as the TransformerEncoder block for most of the args here so this is essentially a swap in replacement. Having each encoder block we ship use different defaults everywhere would be a lot to track, and lead to bad usability.

abheesht17 replied:
Right. I'll set the defaults to the ones used by `TransformerEncoder` 👍🏼

fchollet commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,183 @@
+        intermediate_dim=3072,
+        dropout=0.1,
+        activation="gelu",
+        layer_norm_epsilon=1e-12,
```
This seems way too small

fchollet commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,183 @@
+        dropout=0.1,
+        activation="gelu",
+        layer_norm_epsilon=1e-12,
+        kernel_initializer="glorot_uniform",
```
Is this choice of initializer the best practice for this layer?

fchollet reviewed (commented):
Thanks for the PR!

abheesht17 reviewed (commented):
Thank you for the review comments, @fchollet, @mattdangerw! :)

Summary:

- For all argument-related comments, I've attached links to the config present in the official FNet code. What do you think I should do - keep the same defaults as given in the FNet config, or change it to the defaults given in `TransformerEncoder`?
- Please see the comment on initializers. I've changed everything to `RandomNormal` with a standard deviation of `2e-2`.
- For the comments on making the functions in-line and the one about `build()`, I had followed what's been done in the `TransformerEncoder` class. I've now changed it to what @fchollet mentioned in the comments. Should I make the same changes in `TransformerEncoder` and `TransformerDecoder`?

chenmoneygithub commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,178 @@
+            x = self._output_dense(x)
+            return self._output_dropout(x)
+
+        # Apply fourier transform on the input.
```
I would prefer putting these comments into the class' docstring, since the line `mixing_output = _fourier_transform(inputs)` does not indicate this much information.

Additionally, this code reference is error-prone, because the line number is subject to change. Let's delete it, thanks!

abheesht17 replied:
Done!

abheesht17 reviewed (commented):
Thank you for the comments, @chenmoneygithub! I've addressed them.

mattdangerw commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,177 @@
+            A Tensor of the same shape as the `inputs`.
+        """
+
+        def _fourier_transform(input):
```
I don't think there's any reason to keep this inner functions private with underscores. They are local to the call function.

You only need the underscore for self._x to communicate that it's private.

abheesht17 replied:
Makes sense. Changed!

mattdangerw commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,177 @@
+        # Apply fourier transform on the input.
+        mixing_output = _fourier_transform(inputs)
+
+        # LayerNorm layer.
```
remove the `# LayerNorm layer.` and `# Feedforward layer.` comments, the function names communicate that fine.

abheesht17 replied:
👍🏼

mattdangerw reviewed (commented):
Left some small comments. Once we resolve the default argument discussion this looks good to me.

abheesht17 reviewed (commented):
Thank you, @mattdangerw, for the review! I've changed all defaults to what they are in `TransformerEncoder` (including the initialisers).

mattdangerw reviewed (approved):
Thanks! This looks good to me.

abheesht17 commented:
Thank you, @mattdangerw! 🥳

mattdangerw commented on keras_nlp/layers/fnet_encoder.py:
```diff
@@ -0,0 +1,173 @@
+    Args:
+        intermediate_dim: int, defaults to 3072. The hidden size of feedforward
+            network.
+        dropout: float, defaults to 0.1. The dropout value, applied in the
```
oops just saw this, update the default in the docstrings too (or maybe just drop them), we generate docs we can include the default from the function signature

abheesht17 replied:
Ah, my bad. I've fixed it 👍🏼. Thanks for noticing!

fchollet reviewed (approved):
API LGTM.
